### PR TITLE
[_ASDisplayViewAccessiblity] Improve accessibility support

### PR DESCRIPTION
**Pull request by maicki**
- Improvements for nodes that position is changing dynamically (like in ASTableView / ASCollectionView)
- Performance improvements calculating accessibility elements
- Add compiler flag: ASDK_ACCESSIBILITY_DISABLE to disable custom accessibility code

### CODE REVIEW

File: AsyncDisplayKit/ASDisplayNode.mm [hunk 1]
```diff
@@ -1456,6 +1456,7 @@ - (void)_insertSubnode:(ASDisplayNode *)subnode atSubnodeIndex:(NSInteger)subnod
         // If we can use view API, do. Due to an apple bug, -insertSubview:atIndex: actually wants a LAYER index, which we pass in
         if (canUseViewAPI && sublayerIndex != NSNotFound) {
           [_view insertSubview:subnode.view atIndex:sublayerIndex];
+          [(_ASDisplayView *)_view setAccessibleElements:nil];
```
Comment (appleguy) [inline AsyncDisplayKit/ASDisplayNode.mm]: Should we put these in the view, using:

- (void)didAddSubview:(UIView *)subview;
- (void)willRemoveSubview:(UIView *)subview;

Instead?  Related question, does our current implementation work if the user directly adds a UIView to the hierarchy without a node attached to it?

File: AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm [hunk 2]
```diff
@@ -14,6 +14,27 @@
 
 #pragma mark - UIAccessibilityElement
 
+// Sort accessiblity elements by x and y origin. First sort the elements by y and than by x value.
+static NSArray *SortAccessibilityElements(NSArray *accessibleElements) {
```
Comment (appleguy) [inline AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm]: This is great!  Should we use NSMutableArray?  Less functional, but probably more efficient.  If we did, we'd probably make this function return just the comparator block, and then call the sort method on the mutable array directly.

File: AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm [hunk 3]
```diff
@@ -50,59 +72,84 @@ + (UIAccessibilityElement *)accessibilityElementWithContainer:(id)container node
...
+#pragma mark - UIAccessibility
+
+- (void)setAccessibleElements:(NSArray *)accessibleElements
+{
+  _accessibleElements = nil;
+}
+
+- (NSArray *)accessibleElements
+{
+  ASDisplayNode *viewNode = self.asyncdisplaykit_node;
```
Comment (appleguy) [inline AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm]: Probably just call this node.

File: AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm [hunk 4]
```diff
@@ -112,11 +159,24 @@ - (id)accessibilityElementAtIndex:(NSInteger)index
...
-  return _accessibleElements[index];
+  ASDisplayNode *viewNode = self.asyncdisplaykit_node;
+  if (viewNode == nil) {
+    return nil;
+  }
+
+  // Recalculate the elements
+  NSArray *elements = SortAccessibilityElements(CollectAccessibilityElementsForViewAndNode(self, viewNode));
+
+  // We have to update the frame of the accessiblity element in case it changed
```
Comment (appleguy) [inline AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm]: This solution seems fine.  Just add something like: // the accessibilityFrame is in screen coordinates, so a scroll view moving is a common case where this value needs to change without this element able to detect this (besides performing the coordinate conversation again)

File: AsyncDisplayKit/ASDisplayNode.mm (hunk 1 above)
Comment (maicki) [inline AsyncDisplayKit/ASDisplayNode.mm]: This currently can crash and will be fixed soon

File: AsyncDisplayKit/ASDisplayNode.mm [hunk 5]
```diff
@@ -1634,6 +1635,7 @@ - (void)_addSubnodeSubviewOrSublayer:(ASDisplayNode *)subnode
   BOOL canUseViewAPI = !self.isLayerBacked && !subnode.isLayerBacked;
   if (canUseViewAPI) {
     [_view addSubview:subnode.view];
+    [(_ASDisplayView *)_view setAccessibleElements:nil];
```
Comment (maicki) [inline AsyncDisplayKit/ASDisplayNode.mm]: Fix casting

File: AsyncDisplayKit/ASDisplayNode.mm [hunk 6]
```diff
@@ -1689,6 +1691,7 @@ - (void)removeFromSupernode
         [layer removeFromSuperlayer];
       } else {
         [view removeFromSuperview];
+        [(_ASDisplayView *)_view setAccessibleElements:nil];
```
Comment (maicki) [inline AsyncDisplayKit/ASDisplayNode.mm]: Fix casting

File: AsyncDisplayKit/ASDisplayNode.mm [hunk 7]
```diff
@@ -3149,6 +3152,7 @@ - (void)addSubnode:(ASDisplayNode *)subnode
       [selfNode addSubnode:subnode];
     } else {
       [self addSubview:subnode.view];
+      [(_ASDisplayView *)self setAccessibleElements:nil];
```
Comment (maicki) [inline AsyncDisplayKit/ASDisplayNode.mm]: Fix casting

File: AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm (hunk 2 above)
Comment (Adlai-Holler) [inline AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm]: The comparator block could even be static, using dispatch_once

Comment (maicki) [inline AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm]: I really liked the functional style (and prefer that in general) but it is adding some overhead so we pass in a `NSMutableArray` in all of the collect as well as the sort function.

Furthermore the comparator is static now

File: AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm (hunk 3 above)
Comment (maicki) [inline AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm]: Changed to node

**Comment by maicki** [comment]
@appleguy @Adlai-Holler Ready for review again. Although it's not the final solution we should consider merging it as it's an improvement as well as fixes the order of accessibility elements issue.

File: AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm [hunk 8]
```diff
@@ -32,96 +57,118 @@ + (UIAccessibilityElement *)accessibilityElementWithContainer:(id)container node
...
   ASDisplayNode *viewNode = self.asyncdisplaykit_node;
   if (viewNode == nil) {
-    return nil;
+    return @[];
   }
   
-  // Handle rasterize case
-  if (viewNode.shouldRasterizeDescendants) {
-    _accessibleElements = ASCollectUIAccessibilityElementsForNode(viewNode, viewNode, self);
+  CGRect screenFrame = UIAccessibilityConvertFrameToScreenCoordinates(self.frame, self);
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm]: I believe this should be `self.bounds`. The docs say `The rect should exist in the view space of the UIView argument` and `self.frame` is in the superview's coordinates.

File: AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm [hunk 9]
```diff
@@ -32,96 +57,118 @@ + (UIAccessibilityElement *)accessibilityElementWithContainer:(id)container node
...
+        // No view for layer backed nodes exist. It's necessary to create a UIAccessibilityElement that represents this node
+        accessiblityElement = [UIAccessibilityElement accessibilityElementWithContainer:view node:subnode];
+        
+        CGRect frame = [node convertRect:subnode.bounds fromNode:subnode];
+        [accessiblityElement setAccessibilityFrame:UIAccessibilityConvertFrameToScreenCoordinates(frame, view)];
+        [elements addObject:accessiblityElement];
+      } else {
+        // Accessiblity element is not layer backed just translate the view's frame to the container and add the view
+        accessiblityElement = subnode.view;
+        [accessiblityElement setAccessibilityFrame:UIAccessibilityConvertFrameToScreenCoordinates(subnode.frame, view)];
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm]: Is this necessary? Will the UIView not have the correct accessibilityFrame by default?

File: AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm [hunk 10]
```diff
@@ -32,96 +57,118 @@ + (UIAccessibilityElement *)accessibilityElementWithContainer:(id)container node
...
 }
 
 - (NSInteger)indexOfAccessibilityElement:(id)element
 {
-  if (_accessibleElements == nil) {
-    return NSNotFound;
-  }
-  
-  return [_accessibleElements indexOfObject:element];
+  return [self.accessibleElements indexOfObject:element];
```
Comment (Adlai-Holler) [inline AsyncDisplayKit/Details/_ASDisplayViewAccessiblity.mm]: Let's use `indexOfObjectIdenticalTo:`

**Comment by Adlai-Holler** [comment]
Awesome. Couple comments

**Comment by maicki** [comment]
@Adlai-Holler All comments are addressed. Can I have another look over it. Thanks!

**Comment by Adlai-Holler** [comment]
Great! This stuff isn't exciting but it's important for us to be viable for AAA apps by Google etc

**Pull request merged**